aeron/subscription.go

```diff
@@ -152,30 +153,35 @@ func (sub *Subscription) Poll(handler term.FragmentHandler, fragmentLimit int) i
 			startingIndex = 0
 		}
 
-		for i := startingIndex; i < length && fragmentsRead < fragmentLimit; i++ {
-			fragmentsRead += img[i].Poll(handler, fragmentLimit-fragmentsRead)
+		for i := startingIndex; i < length && fragmentsRead < fragmentLimit && err == nil; i++ {
+			var newFrags int
+			newFrags, err = img[i].Poll(handler, fragmentLimit-fragmentsRead)
+			fragmentsRead += newFrags
```

Comment:
If I were writing this from scratch, definitely.  Since I'm trying to balance existing functionality/expectations, I was a little unsure.  This error bubbles up from FragmentHandler.  From the few that generate errors that I've seen, they call error handlers, but otherwise gracefully continue processing.  I decided to be conservative and preserve existing functionality, which is to keep going in spite of errors.

Yes, it's weird, because the returned error will be the last error returned, hiding any prior ones.  If you think it's safe to newly short-circuit, I'm happy to do that - it definitely feels more natural.

Reply:
What happens in Java/C++ if you throw into the fragment handler?

Reply:
They keep going in spite of errors, as we do here, and it almost looks intentional.  Subscription::poll passes on a fragmentLimit, and if there are no issues, TermReader can hit fragmentLimit in a single call.  But in an error, TermReader calls the error handler, then returns how many fragments were read so far.  Subscription::poll keeps calling TermReader, trying to reach fragmentLimit no matter how many calls/errors it takes.

https://github.com/real-logic/aeron/blob/3a7c94d0825852d9f5ae75b6828a7302f51fccea/aeron-client/src/main/java/io/aeron/logbuffer/TermReader.java#L82

https://github.com/real-logic/aeron/blob/3a7c94d0825852d9f5ae75b6828a7302f51fccea/aeron-client/src/main/java/io/aeron/Subscription.java#L193

By switching to an error-return vs. error-handler model, we only get to see one error vs. they have the opportunity to see them all.  If we short circuited on each error, the caller has the option to continue polling or short circuit themselves based on the nature of the error.  But callers would need to newly implement that code, vs. relying on it always continuing (as it currently does, with or without this PR).

I defer to you on that design decision.  I see arguments either way.  I'd be curious if the errors in practice are actually actionable, or if the error handler is a black hole in reality, or if errors are so rare that it doesn't matter what decision we make.  But I believe fragment handlers are proprietary code generally, so I don't have visibility into that.

Reply:
The Java code makes me think we shouldn't let you return an error into the handler, and we should back out the fragmenthandler change. If it's just going to call an error handler anyway. In Go, we can just not allow you to return an error and force the user to do the error handling themselves, which looks like what Java wants you to do anyway.

Reply:
Sounds good.  I backed out that change.